### ISSUE TITLE
Can't roundtrip FSharp.Core.dll

### ISSUE THREAD

**Issue by vbfox** [jbevain/cecil#215]
Using Cecil `FSharp.Core dll` can't be loaded and saved again (At least for version 4.3.1 of the dll)

The code throw `NotImplementedException` with a stacktrace of : 

```
   at Mono.Cecil.SignatureReader.GetPrimitiveType(ElementType etype)
   at Mono.Cecil.SignatureReader.ReadTypeSignature(ElementType etype)
   at Mono.Cecil.SignatureReader.ReadMethodSignature(IMethodSignature method)
   at Mono.Cecil.MetadataReader.ReadMemberReferenceSignature(UInt32 signature, TypeReference declaring_type)
   at Mono.Cecil.MetadataReader.ReadTypeMemberReference(MetadataToken type, String name, UInt32 signature)
   at Mono.Cecil.MetadataReader.ReadMemberReference(UInt32 rid)
   at Mono.Cecil.MetadataReader.GetMemberReference(UInt32 rid)
   at Mono.Cecil.MetadataReader.LookupToken(MetadataToken token)
   at Mono.Cecil.Cil.CodeReader.PatchRawCode(ByteBuffer buffer, Int32 code_size, CodeWriter writer)
   at Mono.Cecil.Cil.CodeReader.PatchRawFatMethod(ByteBuffer buffer, MethodSymbols symbols, CodeWriter writer, MetadataToken& local_var_token)
   at Mono.Cecil.Cil.CodeReader.PatchRawMethodBody(MethodDefinition method, CodeWriter writer, MethodSymbols& symbols)
   at Mono.Cecil.Cil.CodeWriter.WriteUnresolvedMethodBody(MethodDefinition method)
   at Mono.Cecil.Cil.CodeWriter.WriteMethodBody(MethodDefinition method)
   at Mono.Cecil.MetadataBuilder.AddMethod(MethodDefinition method)
   at Mono.Cecil.MetadataBuilder.AddMethods(TypeDefinition type)
   at Mono.Cecil.MetadataBuilder.AddType(TypeDefinition type)
   at Mono.Cecil.MetadataBuilder.AddTypeDefs()
   at Mono.Cecil.MetadataBuilder.BuildTypes()
   at Mono.Cecil.MetadataBuilder.BuildModule()
   at Mono.Cecil.ModuleWriter.<BuildMetadata>b__0(MetadataBuilder builder, MetadataReader _)
   at Mono.Cecil.ModuleDefinition.Read[TItem,TRet](TItem item, Func`3 read)
   at Mono.Cecil.ModuleWriter.BuildMetadata(ModuleDefinition module, MetadataBuilder metadata)
   at Mono.Cecil.ModuleWriter.WriteModuleTo(ModuleDefinition module, Stream stream, WriterParameters parameters)
   at Mono.Cecil.ModuleDefinition.Write(Stream stream, WriterParameters parameters)
   at Mono.Cecil.ModuleDefinition.Write(String fileName, WriterParameters parameters)
   at Mono.Cecil.AssemblyDefinition.Write(String fileName)
```

The problem seem to be due to the usage of `methodhandleof ` in https://github.com/fsharp/fsharp/blob/master/src/fsharp/FSharp.Core/Query.fs#L363 generating code looking like that in ILDasm :

```
  IL_000b:  ldtoken    method class [System.Core]System.Linq.Expressions.Expression`1<!!0> Microsoft.FSharp.Linq.RuntimeHelpers.LeafExpressionConverter::ImplicitExpressionConversionHelper(!!0)
  IL_0010:  call       class [mscorlib]System.Reflection.MethodBase [mscorlib]System.Reflection.MethodBase::GetMethodFromHandle(valuetype [mscorlib]System.RuntimeMethodHandle)
```

The current code seem to have problem with that because the type of the parameter is `ElementType.None` and there is no correct value for `SignatureReader.GetPrimitiveType` to map it to.

**Comment by jbevain** [comment]
Hey @vbfox, is this with 0.9.6 or on master?

**Comment by jbevain** [comment]
Also, can you upload the DLL somewhere so I'm sure we're talking about the same one?

**Comment by vbfox** [comment]
I tried the NuGet version and was master I think.
But I also tried the version from https://github.com/gluck/il-repack so i'll need to double-check.

I'll try with a clean build from master tonight and upload a test project if I reproduce.

**Comment by jbevain** [comment]
Yeah I can't repro here with the current master de60c2a129dacdec0f09fdf3e97b767fe819b6b3 and FSharp.Core.dll 4.3.1 from my GAC.

**Comment by vbfox** [comment]
Verry sorry to have bothered you.

I checked against master and can't reproduce all `FSharp.Core.dll` that I have in my GAC roundtrip correctly.

As I loaded all assemblies from a folder I think that I forgot to double check the assembly file creating the problem, it was in FSharp.Core code, but not in the real dll. It was in a version merged via ILMerge with other assemblies.
And it seem that ILMerge failed somewhere (not Cecil) as the same assembly also fail PEVerify.
